### ISSUE TITLE
[Panel] General UI improvements

### ISSUE THREAD

**Issue by mehcode** [steelbrain/linter-ui-default#352]
@steelbrain How willing are we to change the panel? I see a lot of work has gone into the current iteration of the panel. 

To be blunt, it's not very pretty. I like pretty so I'm going to invest some time to see if I can help make it pretty. 

---

Some ideas (not particularly novel but still) that I have:

 - Group the table by filename (with the filename on its own line with a chevron to expand/close the second). This would give us room to add a couple of useful pieces of information to the table.
    - File path could be added after filename
    - File *icon* could be added for visual grep
    - Badge (with per-file issue count) could be added. This would obviate #234 as per-file count would be easy to see.

 - *Always* show all issues in project in the panel (grouped by file) and focus the current file when switching files (make it a setting if we have to).

 - The words `Warning` and `Error` are quite verbose. A colored icon copied from the status bar would save a bunch of space.

 - Why did we decide on a table / column format? Through grouping and icons (discussed above) we can drop 2 of the columns. We could easily do `Error message here (line:column)` and use a faded color for the `line:column` and drop that column. The only one left is then an optional column that could be folded into the description column using prefixed labels (though not colored labels like we had before). I get that columns let you sort it but I don't really think sorting adds much utility. If I was going to ask for anything to help manage my list of issues I'd want filtering.

---

I realize you might just tell me to go away and write my own linter-ui package. I get that. If I must,  I will, I'd just *greatly* prefer to have us agree to a solid UX in the default package.

**Comment by steelbrain** [comment]
I'm willing to discuss any chances to the Panel, no matter how drastic if they are worth the effort. The current implementation is not very pretty as you noted but I've personally found it to be more useful than the previous UI.

The reason I went with tables is visual consistency, previously each error had a different message and each message usually had a different length so you'd have to read the whole line just to get the `on /path/to/file` part, that is now much much easier to read because everything's in their appropriate boxes, you couldn't sort things like you can with one or more columns like you can do now.

Through my experience of managing these Linter and Linter UI projects (which are by far my most popular open source projects by number of users so far), I've concluded that more options doesn't always mean better, the latest linter doesn't even have half the options of the previous major version, too much configuration ended me up with a codebase that was hard to maintain and a set of configuration that was only rarely useful and removing them would upset people because it broke their workflow, kinda like https://xkcd.com/1172/

Writing your own UI package isn't a bad idea at all, I would actually encourage you to do so and if yours looks better than this one and is easier to operate, we can always steal the code :)

**Comment by mehcode** [comment]
> I'm willing to discuss any chances to the Panel, no matter how drastic if they are worth the effort. The current implementation is not very pretty as you noted but I've personally found it to be more useful than the previous UI.

I've compared the previous UI and the current UI. I'd argue the previous UI was a touch prettier but the newer one is definitely more elegant.

> Through my experience of managing these Linter and Linter UI projects (which are by far my most popular open source projects by number of users so far), I've concluded that more options doesn't always mean better [...]

Totally agree with you. The number of options the linter-ui has _now_ is a bit intense. I didn't realize it had more before.

---

> Writing your own UI package isn't a bad idea at all, I would actually encourage you to do so and if yours looks better than this one and is easier to operate, we can always steal the code :)

I think this is what i'll do. Some of the changes I'd like to do are pretty drastic and it's always more fun to start with a clean slate and then move things in vs. adjusting what's there. 

Happy to see you're open to changes. I'll look forward to showing off what I can come up with.

**Comment by dmmulroy** [comment]
This seemed like the right place to bring this up instead of opening up a new issue, but I'd love to have setting for the bottom panel that enables/disables it from auto opening when lint issues occur.

**Comment by Arcanemagus** [comment]
@dmmulroy You mean the "Hide Panel When Empty" setting [already there](https://github.com/steelbrain/linter-ui-default/blob/v1.6.2/package.json#L98-L103)? 😛

**Comment by dmmulroy** [comment]
@Arcanemagus The way that setting reads is that it only closes the panel if there are no issues.  I want the opposite, I want the panel to stay closed even if there are lint issues.  Maybe that's what this setting does and I'm misunderstanding it haha

**Comment by Arcanemagus** [comment]
No, you're understanding it correctly, I just thought you meant that. Please file a new issue about that for further discussion.

**Comment by mehcode** [comment]
@steelbrain @Arcanemagus 

Finally got some time to do this. Here is what I'm working on (a new panel):

![screen shot 2017-06-30 at 9 58 48 pm](https://user-images.githubusercontent.com/753919/27759196-543eb3d4-5ddf-11e7-8881-e9acfdd505bc.png)

---

For reference here are some other linters

 - [Atom Linter v2](https://user-images.githubusercontent.com/753919/27759182-0673865c-5ddf-11e7-9b34-eb0c76713077.png)
 - [Atom Linter v1](https://user-images.githubusercontent.com/753919/27759192-40868858-5ddf-11e7-9610-9c4da55fa3e9.png)
 - [VS Code](https://user-images.githubusercontent.com/753919/27759223-d06dff0a-5ddf-11e7-9ff5-021607be0ff3.png)
 - [Xcode](http://i.stack.imgur.com/Xf14y.png)
 - [Sublime Text 3](https://user-images.githubusercontent.com/753919/27759958-0ee8879e-5df1-11e7-8e96-379f77c64604.png)
 - [Eclipse](https://images.duckduckgo.com/iu/?u=https%3A%2F%2Fwww.eclipse.org%2Fforums%2Findex.php%2Ffa%2F6135%2F0%2F&f=1)
 - [Vim](https://cloud.githubusercontent.com/assets/2831985/24484590/d321bb6c-14b3-11e7-89a2-ae2c74a9f52d.png)




---

Tell me what you think of what I've got so far. It's definitely just a first draft. I think VS Code (just talking about the panel) gets the most _right_ for UX out of the existing implementations but there is room to improve. The following are also ideas, not final.
 
 - Show the provider 
 - A way to filter all / errors / warnings / infos. Perhaps a toggle group in the top-left of the panel.
 - A way to search through messages. For instance, the rule name is in the message for eslint so searching "semi" would show all the semi warnings. Could also make the search a touch expressive with `file:...` and `provider:...` expressions.
 - Sorting? I've been thinking on the utility of it but I can't imagine a use case for this. @steelbrain I'm curious why it was added to this version. Line/File sort only makes sense (to me) if you are doing a sort  for `(file, line)` as a pair. Severity sorting could make sense but I feel severity _toggling_ covers the use case there even better. And provider filtering makes more sense than provider sorting.
 - Fixing / Resolutions should definitely be here somewhere. 
 - As I finish writing this I'm thinking I should move the line/column to the left of the severity symbol and use the same format as the status bar.

Some more thoughts

- Linting should be project-wide all the time. File-specific at a panel, status bar, or tree-view context makes little UX sense (and the rest of the editor world agrees with me). @steelbrain What's up with linter-eslint? Why can `eslint .` finish instantly on my command line for my whole project but `linter-eslint` not show me all my errors in my project? I realize it's probably just an implementation leftover from atom linter not originally doing project-wide (iirc) , I'm just saying we should make a push for all linter packages here. It's confusing otherwise.

 - Most linter implementations in editors tend to have a panel be very "tack on" (or missing) where as IDEs tend to overshoot and have way too much going on. Just interesting.

 - I *really* don't like the centered word "Linter" in the dock/tab/thing. It looks great on the left/right docks but the bottom dock looks really odd. If I get some more time I'll write something up for @simurai on the .. not really sure where to put that issue so I guess I'll put in Atom.

---

Sorry for how long this post got. I really should learn how not to ramble so much. Props to @steelbrain and @Arcanemagus for how awesome the linter service is. 

https://github.com/mehcode/linter-ui-plus

**Comment by Positive07** [comment]
I would like to see the linter name somewhere, is that something you are considering?

Like for example I use TypeScript and TSLint on the same code, but TypeScript errors are far more important to me... I would like to at least differentiate them.

On another note, is your package on feature parity with linter-ui-default regarding the non-dock aspects like tooltips and such? if so, do you think it's ready to install?

**Comment by Arcanemagus** [comment]
> Show the provider

Yes please! I usually have 2-3 providers for the file types I work in. @Positive07 this is what you mean by "linter name" btw 😉.

> Sorting? I've been thinking on the utility of it but I can't imagine a use case for this. @steelbrain I'm curious why it was added to this version.

I use it all the time myself... but I have Panel set to File instead of Project so it's more limited in scope than the project wide one you are working on.

> What's up with linter-eslint ... not show[ing] me all my errors in my project?

This is actually _far_ more complex than you'd think at a glance. `linter-eslint` uses the Node.js API, and as such would need to run it on every single file manually which would definitely not be the "instantly" that you are expecting from the CLI. There are a lot of major changes that would need to be made and incredible complexity added there on top of the already crazy complexity it has to do anything like that.

> I'm just saying we should make a push for all linter packages here. [project wide linting]

It's actually _very_ rare to find a project wide linter, `flow` is only of the few ones that I know of. Although ESLint can run on a directory in the CLI, it isn't actually a project wide linter either.

> I really don't like the centered word "Linter" in the dock/tab/thing.

This is actually just a function of Docks and how your UI theme works with them. I'm guessing you are using One Dark in the screenshots there. Atom Material for example has them looking like file tabs.

> Props to @steelbrain and @Arcanemagus for how awesome the linter service is.

@steelbrain Has done the UI here, I just work on the providers (and complain about what's broken in the UI that I notice 😛).

**Comment by Positive07** [comment]
> Show the provider

> Yes please! I usually have 2-3 providers for the file types I work in. @Positive07 this is what you mean by "linter name" btw 😉.

That's totally what I mean, please do

**Comment by mehcode** [comment]
![screen shot 2017-07-03 at 4 14 37 pm](https://user-images.githubusercontent.com/753919/27810073-cc913686-600a-11e7-99a3-925eaf429274.png)
![screen shot 2017-07-03 at 4 13 45 pm](https://user-images.githubusercontent.com/753919/27810074-cca38b9c-600a-11e7-9d83-0d2a34c01006.png)
![screen shot 2017-07-03 at 4 12 49 pm](https://user-images.githubusercontent.com/753919/27810075-ccb4f40e-600a-11e7-89eb-ccc66f0ed3d9.png)

Some options for the provider name. I think I'm leaning towards the last one.

--- 


> This is actually far more complex than you'd think at a glance. linter-eslint uses the Node.js API, 

Change https://github.com/AtomLinter/linter-eslint/blob/master/src/worker.js#L14 to use `engine.executeOnFiles(".")` and 💥 

Found that out with some poking in nesh

**Comment by steelbrain** [comment]
@mehcode No lint on change that way, only lint on save 😉

**Comment by Positive07** [comment]
> Some options for the provider name. I think I'm leaning towards the last one.

So hard to choose, the first two look great,  I don't really like the third one but it could be useful to easily spot the different linters... I really don't know

**Comment by Arcanemagus** [comment]
What about putting the provider name first?

**Comment by adamreisnz** [comment]
+1 for having color coded icons, see #383

**Comment by AshlessCinder** [comment]
> Some options for the provider name. I think I'm leaning towards the last one.

I'd like to throw a vote in for the first option.
It distracts the least from the important content (the actual error), but it's still clear and easy to see if you're looking for what provider threw the error.

Your screenshots look amazing. Keep up the good work :D

**Comment by Arcanemagus** [comment]
@AshlessCinder You can try out @mehcode's version over on https://github.com/mehcode/linter-ui-plus, that's also where comments, suggestions, and bugs with that implementation should go 😉.
